src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,562 @@
+                lr = tf.exp(
+                    ll - self._old_policy.distribution.log_prob(i.action_var))
+                surrogate = lr * adv
+
+            # Finalize objective function
+            with tf.name_scope('loss'):
+                if self._pg_loss == 'vanilla':
+                    # VPG uses the vanilla objective
+                    obj = tf.identity(vanilla, name='vanilla_obj')
+                elif self._pg_loss == 'surrogate':
+                    # TRPO uses the standard surrogate objective
+                    obj = tf.identity(surrogate, name='surr_obj')
+                elif self._pg_loss == 'surrogate_clip':
+                    lr_clip = tf.clip_by_value(lr,
+                                               1 - self._lr_clip_range,
+                                               1 + self._lr_clip_range,
+                                               name='lr_clip')
+                    surr_clip = lr_clip * adv
+                    obj = tf.minimum(surrogate, surr_clip, name='surr_obj')
+
+                if self._entropy_regularzied:
+                    obj += self._policy_ent_coeff * policy_entropy
+
+                # Maximize E[surrogate objective] by minimizing
+                # -E_t[surrogate objective]
+                if self.policy.recurrent:
+                    # filter valid obj
+                    obj = tf.boolean_mask(obj, i.valid_var)
+                loss = -tf.reduce_mean(obj)
+
+            # Diagnostic functions
+            self._f_policy_kl = compile_function(
+                flatten_inputs(self._policy_opt_inputs), pol_mean_kl)
+
+            self._f_rewards = compile_function(
+                flatten_inputs(self._policy_opt_inputs), rewards)
+
+            returns = discounted_returns(self.discount, self.max_path_length,
+                                         rewards)
+            self._f_returns = compile_function(
+                flatten_inputs(self._policy_opt_inputs), returns)
+
+            return loss, pol_mean_kl
+
+    def _build_entropy_term(self, i):
+        """Build policy entropy tensor.
+
+        Args:
+            i (namedtuple): Collection of variables to compute policy loss.
+
+        Returns:
+            tf.Tensor: Policy entropy.
+
+        """
+        pol_dist = self.policy.distribution
+
+        with tf.name_scope('policy_entropy'):
+            if self._use_neg_logli_entropy:
+                policy_entropy = -pol_dist.log_prob(i.action_var,
+                                                    name='policy_log_likeli')
+            else:
+                policy_entropy = pol_dist.entropy()
+
+            # This prevents entropy from becoming negative for small policy std
+            if self._use_softplus_entropy:
+                policy_entropy = tf.nn.softplus(policy_entropy)
+
+            if self._stop_entropy_gradient:
+                policy_entropy = tf.stop_gradient(policy_entropy)
+
+        if not self.policy.recurrent:
```

Comment:
would this still work if the policy were non-recurrent?

Reply:
We have tests and benchmarks for NPO with non-recurrent policies.

Reply:
i'm asking if we can do the same thing for both recurrent and non-recurrent here